editor/store-persist.js

```diff
@@ -7,13 +12,14 @@ const DEFAULT_STORAGE_KEY = 'REDUX_PERSIST';
 
 /**
  * Store enhancer to persist a specified reducer key
- * @param {String}     reducerKey The reducer key to persist
- * @param {String}     storageKey The storage key to use
- * @param {Object}     defaults   Default values
+ * @param {String}     reducerKey          The reducer key to persist
+ * @param {String}     storageKey          The storage key to use
+ * @param {Object}     defaults            Default values
+ * @param {Object}     beforeReHydrate     Preprocess of the payload before rehydrate.
  *
  * @return {Function}             Store enhancer
  */
-export default function storePersist( reducerKey, storageKey = DEFAULT_STORAGE_KEY, defaults = STORE_DEFAULTS ) {
+export default function storePersist( reducerKey, storageKey = DEFAULT_STORAGE_KEY, defaults = STORE_DEFAULTS, beforeReHydrate = identity ) {
```

Comment:
Aside from making the argument signature a bit more awkward, it's not clear to me _why_ we need `defaults`, and particularly why it needs to be hard-coded to editor-specific defaults. In the `persistedState` assignment below, wouldn't `store.getState()` already reflect these defaults?

Mentioned because we're importing `STORE_DEFAULTS` now in three places: the enhancer (which arguably should not be aware of these defaults itself), editor initialization, and the store reducer. It seems like only one of the last two should care about this. We already have the reducer using defaults (e.g. preferences), _or_ [Redux's `createStore`](http://redux.js.org/docs/api/createStore.html) allows for a preloaded state as the second argument.

Also mentioning because as we add more arguments here, particularly when the arguments have defaults and the ordering isn't important, it seems like we might consider refactoring arguments to an object of options, e.g.

```js
storePersist( {
	reducerKey: '...',
	storageKey: '...',
	beforeRehydrate: () => {},
} )
```

Also helps lend readability to the invocation, since without intimate familiarity with the utility, it is not clear what each of the arguments below is meant for:

```js
storePersist( 'preferences', GUTENBERG_PREFERENCES_KEY, STORE_DEFAULTS, disableIsSidebarOpenedOnMobile ),
```

Reply:
Hi @aduth, I also think using STORE_DEFAULTS in reducer is enough. And my tests seem to confirm that. So I applied the change and only the reducer sets the defaults. I also used an options object in storePersist. In order for the enhancer to not include editor specific details, I added a parameter storePersistOptions and now is the provider when creating the store that passes details, still not the perfect solution but may be an improvement.

Reply:
>In order for the enhancer to not include editor specific details, I added a parameter storePersistOptions and now is the provider when creating the store that passes details, still not the perfect solution but may be an improvement.

As I see it, it's fine for details of `editor/store.js` to be editor-specific, particularly since it is responsible for bringing in and defining the editor's reducer/effects behaviors. The benefit of making the persistence enhancer editor-unaware is if we choose to make it available as a general-purpose utility, and arguably helps retain a nice separation of concerns. There isn't as obvious a benefit in separating the editor and its store initialization.

Reply:
> As I see it, it's fine for details of editor/store.js to be editor-specific, particularly since it is responsible for bringing in and defining the editor's reducer/effects behaviors.

Hi @aduth, I agree with you, with that in mind I think the current changes respect this. I moved editor specific defaults out of store persist and they are passed in editor/store.js.

Reply:
> As I see it, it's fine for details of editor/store.js to be editor-specific, particularly since it is responsible for bringing in and defining the editor's reducer/effects behaviors.

Hi @aduth, I agree with this, with that in mind I removed the editor specific details that were already in editor/store-persist.js and moved them into editor/store.js.